cli/command_repository_upgrade.go

```diff
@@ -0,0 +1,291 @@
+		Validate(func(tmpCmd *kingpin.CmdClause) error {
+			if v := os.Getenv(c.svc.EnvName(upgradeLockFeatureEnv)); v == "" {
+				return errors.Errorf("please set %q env variable to use this feature", upgradeLockFeatureEnv)
+			}
+			return nil
+		})
+
+	beginCmd := parent.Command("begin", "Begin upgrade.").Default()
+	beginCmd.Flag("advance-notice", "Advance notice for upgrade to allow enough time for other Kopia clients to notice the lock").DurationVar(&c.advanceNoticeDuration)
+	beginCmd.Flag("io-drain-timeout", "Max time it should take all other Kopia clients to drop repository connections").Default(repo.DefaultRepositoryBlobCacheDuration.String()).DurationVar(&c.ioDrainTimeout)
+	beginCmd.Flag("force", "Force using an unsafe io-drain-timeout").Default("false").Hidden().BoolVar(&c.force)
+	beginCmd.Flag("status-poll-interval", "An advisory polling interval to check for the status of upgrade").Default("60s").DurationVar(&c.statusPollInterval)
+
+	// upgrade phases
+
+	// Set the upgrade lock intent.
+	beginCmd.Action(svc.directRepositoryWriteAction(c.runPhase(c.setLockIntent)))
+	// If requested then drain all the clients otherwise stop here.
+	beginCmd.Action(svc.directRepositoryWriteAction(c.runPhase(c.drainOrCommit)))
+	// If the lock is fully established then perform the upgrade.
+	beginCmd.Action(svc.directRepositoryWriteAction(c.runPhase(c.upgrade)))
+	// Commit the upgrade and revoke the lock, this will also cleanup any
+	// backups used for rollback.
+	beginCmd.Action(svc.directRepositoryWriteAction(c.runPhase(c.commitUpgrade)))
+
+	rollbackCmd := parent.Command("rollback", "Rollback the repository upgrade.")
+	rollbackCmd.Flag("force", "Force rollback the repository upgrade, this action can cause repository corruption").BoolVar(&c.forceRollback)
+
+	rollbackCmd.Action(svc.directRepositoryWriteAction(c.forceRollbackAction))
+
+	c.svc = svc
+}
+
+func (c *commandRepositoryUpgrade) forceRollbackAction(ctx context.Context, rep repo.DirectRepositoryWriter) error {
+	if !c.forceRollback {
+		return errors.New("repository upgrade lock can only be revoked unsafely; please use the --force flag")
+	}
+
+	if err := rep.RollbackUpgrade(ctx); err != nil {
+		return errors.Wrap(err, "failed to rollback the upgrade")
+	}
+
+	log(ctx).Infof("Repository upgrade lock has been revoked.")
+
+	return nil
+}
+
+func (c *commandRepositoryUpgrade) runPhase(act func(context.Context, repo.DirectRepositoryWriter) error) func(context.Context, repo.DirectRepositoryWriter) error {
+	return func(ctx context.Context, rep repo.DirectRepositoryWriter) error {
+		if !c.skip {
+			if err := act(ctx, rep); err != nil {
+				// Explicitly skip all stages on error because tests do not
+				// skip/exit on error. Tests override os.Exit() that prevents
+				// running rest of the phases until we set the skip flag here.
+				// This flag is designed for testability and also to support
+				// rollback.
+				c.skip = true
+				return err
+			}
+		}
+
+		return nil
```

Comment:
Invert the check here to simplify these blocks?

```suggestion
		if c.skip {
			return nil
		}
		if err := act(ctx, rep); err != nil {
			// Explicitly skip all stages on error because tests do not
			// skip/exit on error. Tests override os.Exit() that prevents
			// running rest of the phases until we set the skip flag here.
			// This flag is designed for testability and also to support
			// rollback.
			c.skip = true
			return err
		}

```

Reply:
thank you! committed